contributors/design-proposals/api-machinery/customresources-versioning.md

```diff
@@ -0,0 +1,177 @@
+  // Names are the names used to describe this custom resource
+  Names CustomResourceDefinitionNames
+  // Scope indicates whether this resource is cluster or namespace scoped.  Default is namespaced
+  Scope ResourceScope
+  // Validation describes the validation methods for CustomResources
+  Validation *CustomResourceValidation
+
+  // ***************
+  // ** New Field **
+  // ***************
+  // Versions is the list of all supported versions for this resource.
+  // Validation: All versions must use the same validation schema for now. i.e., top 
+  // level Validation field is applied to all of these versions.
+  // Order: The order of these versions is used to determine the order in discovery API
+  // (preferred version first).
+  // The versions in this list may not be removed if they are in 
+  // CustomResourceDefinitionStatus.StoredVersions list.
+  Versions []CustomResourceDefinitionVersion
+}
+
+// ***************
+// ** New Type **
+// ***************
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+}
+```
+
+The Status object will have a list of potential stored objects. This data is necessary to do a storage migration in future (the author can choose to do the migration themselves but there is a plan [TODO(lavalamp): link?] to solve the problem of migration, potentially for both standard and custom types).
+
+```golang
+// CustomResourceDefinitionStatus indicates the state of the CustomResourceDefinition
+type CustomResourceDefinitionStatus struct {
+  ...
+
+  // StoredVersions are all versions ever marked as storage in spec. Tracking these
+  // versions allow a migration path for stored version in etcd. The field is mutable
+  // so the migration controller can first make sure a version is certified (i.e. all
+  // stored objects is that version) then remove the rest of the versions from this list.
+  // None of the versions in this list can be removed from the spec.Versions field.
+  StoredVersions []string
+}
+```
+
+Basic validations needed for the `version` field are:
+
+* `Spec.Version` field exists in `Spec.Versions` field (serve should be True for that version if there is any).
+* If there is any `Served` `Version` in list of `Versions`, then the `Spec.Version`’s version in the `Spec.Versions` list should have a set `Served` flag. This is for backward compatibility. An old controller expect that version to be served but only the whole CRD is served. CRD Registration controller should unregister a CRD with no serving version.
+* None of the `Status.StoredVersion` can be removed from `Spec.Versions` list.
+* Only one of the versions in `spec.Versions` can flag as `Storage` version.
+
+# **Declarative conversion**
```

Comment:
Ah, ok.  I think I see what happened.  You're saying "these are some thoughts I had", not "I want to build this an merge it in 1.11".  If that's the case, can you add a "Declarative conversion design not approved in this document" in big bold letters leading the section?

Reply:
Actually these are the things I want to have in 1.11 (only rename). Here we can discuss it if you have any thoughts. These are designed I discussed in one form or another with Stephan, Daniel, and Eric but none approved it yet. I guess that is the goal of this PR.